### ISSUE TITLE
Rule proposal: Prepend type variables in unions: `A | {B: string}` instead of `{B: string} | A`

### ISSUE THREAD

**Issue by fregante** [sindresorhus/eslint-plugin-unicorn#1758]
### Description

Large inline type definitions may shadow trailing unions with pre-existing types. I think that there was something similar with regular javascript objects (I thought it was [object-shorthand](https://eslint.org/docs/rules/object-shorthand))

```js
const foo = {
 	bar, // shorthand first
	foo: 1 // inline values after
}
```

This is unrelated to my request though, which is about types.

### Fail

```ts
type ElementUnion = {
    readonly attributes: NamedNodeMap;
    /** Allows for manipulation of element's class content attribute as a set of whitespace-separated tokens through a DOMTokenList object. */
    readonly classList: DOMTokenList;
    /** Returns the value of element's class content attribute. Can be set to change it. */
    className: string;
    readonly clientHeight: number;
    readonly clientLeft: number;
    readonly clientTop: number;
    readonly clientWidth: number;
} | Other;
```
```ts
type ElementIntersection = {
    readonly clientHeight: number;
    readonly clientLeft: number;
    readonly clientTop: number;
    readonly clientWidth: number;
} & Other;
```

### Pass

```ts
type ElementUnion = Other | {
    readonly attributes: NamedNodeMap;
    /** Allows for manipulation of element's class content attribute as a set of whitespace-separated tokens through a DOMTokenList object. */
    readonly classList: DOMTokenList;
    /** Returns the value of element's class content attribute. Can be set to change it. */
    className: string;
    readonly clientHeight: number;
    readonly clientLeft: number;
    readonly clientTop: number;
    readonly clientWidth: number;
};
```
```ts
type ElementIntersection = Other & {
    readonly clientHeight: number;
    readonly clientLeft: number;
    readonly clientTop: number;
    readonly clientWidth: number;
};
```

**Comment by sindresorhus** [comment]
Any suggestions for a rule name?

**Comment by fregante** [comment]
`short-values-first`
`short-types-first`
`identifiers-first`

**Comment by papb** [comment]
> Large inline type definitions may shadow trailing unions with pre-existing types.

Hello, can you clarify what you mean by _shadow_? Are you referring just to readability, or something else?

**Comment by fregante** [comment]
Yes, readability

**Comment by fregante** [comment]
> Maybe you should try opening this on [typescript-eslint/typescript-eslint](https://github.com/typescript-eslint/typescript-eslint) first?

I don’t think they accept style rules anymore: https://github.com/typescript-eslint/typescript-eslint/issues/5710#issuecomment-3697280225
